packages/rpc-graphql/src/resolvers/resolve-info/visitor.ts

```diff
@@ -0,0 +1,95 @@
+import {
+    ASTNode,
+    ASTVisitFn,
+    BREAK,
+    FieldNode,
+    FragmentDefinitionNode,
+    FragmentSpreadNode,
+    getNamedType,
+    GraphQLResolveInfo,
+    InlineFragmentNode,
+    visit
+} from 'graphql';
+
+type InjectableVisitorOperations = {
+    fieldNodeOperation: (
+        info: GraphQLResolveInfo,
+        ...args: Parameters<ASTVisitFn<FieldNode>>
+    ) => void;
+    fragmentSpreadNodeOperation: (
+        info: GraphQLResolveInfo,
+        fragment: FragmentDefinitionNode,
+        ...args: Parameters<ASTVisitFn<FragmentSpreadNode>>
+    ) => void;
+    inlineFragmentNodeOperation: (
+        info: GraphQLResolveInfo,
+        ...args: Parameters<ASTVisitFn<InlineFragmentNode>>
+    ) => void;
+}
+type RootNode = FieldNode | FragmentDefinitionNode | InlineFragmentNode;
+
+/**
+ * An AST visitor that keys on the root field provided.
+ * This visitor can be injected with custom logic for various types of nodes.
+ */
+export function injectableRootVisitor(
+    info: GraphQLResolveInfo,
+    rootNode: RootNode | null,
+    operations: InjectableVisitorOperations,
+) {
+    const { fieldNodes, schema } = info;
+    const root = rootNode ?? fieldNodes[0];
+    const parentIsRoot = (ancestors: readonly (ASTNode | readonly ASTNode[])[]) =>
+        (Array.isArray(ancestors) ? ancestors[0] : ancestors) === root;
+    visit(root, {
+        Field(node, key, parent, path, ancestors) {
+            if (!parentIsRoot(ancestors)) return;
+            return operations.fieldNodeOperation(info, node, key, parent, path, ancestors);
+        },
+        FragmentSpread(node, key, parent, path, ancestors) {
+            const fragmentDefinition = info.fragments[node.name.value];
+            const fragmentType = schema.getType(fragmentDefinition.typeCondition.name.value)?.name;
+            let rootType;
+            if (root.kind === 'Field') {
```

Comment:
God I wish this thing let you write AST _selectors_. https://medium.com/dailyjs/eslint-selectors-just-got-super-powers-402ab942def5

Then you could do things like:

```ts
visit(node, {
    'FragmentSpread > Field': (node, key, parent, path, ancestors) => {
        // Only finds `Field` nodes who are the children of `FragmentSpread` nodes.
    },
});
```

If only there were like a TSQuery for GraphQL? https://medium.com/@phenomnominal/easier-typescript-tooling-with-tsquery-d74f04f2b29d